### DISCUSSION TITLE
Custom casting example

### DISCUSSION THREAD

**Discussion by mrgpd-pfevre** [ozlerhakan/poiji#252] [Q&A] [answered]
Hi there,

First I'd like to send my best regards to contributors of this project.
It's amazing and pleasant to use.
I have one question regarding custom casting.
can you provide an example on how to implement, I have actually to deal with a Boolean property, and in excel it can be Yes/No, true/false (that work) and in french oui/non

Best Regards

**Comment by ozlerhakan** [comment]
Hi @mrgpd-pfevre ,

Thank you first off :) have you tried the steps of Custom Casting Implementation writen in README?

**Comment by mrgpd-pfevre** [comment]
Yes I have read the doc but the chapter is a bit short, and I miss an example on how to determine on which cell the cast class will apply.

Pierre Fèvre
Directeur Technique
[logo]<https://www.mission-rgpd.com/>
[phone icon] +33 6 12 69 13 67
[mail icon] ***@***.******@***.***>
[website icon] www.mission-rgpd.com<https://mission-rgpd.com>
[linkedin icon] @missionrgpd<https://www.linkedin.com/company/missionrgpd/>
[banner]<https://www.mission-rgpd.com/demo-logiciel-rgpd/>


De : Hakan Özler ***@***.***>
Envoyé : jeudi 11 août 2022 11:26
À : ozlerhakan/poiji ***@***.***>
Cc : Pierre FEVRE ***@***.***>; Mention ***@***.***>
Objet : Re: [ozlerhakan/poiji] Custom casting example (Discussion #252)


Hi @mrgpd-pfevre<https://github.com/mrgpd-pfevre> ,

Thank you first off :) have you tried the steps of Custom Casting Implementation writen in README?

**Comment by ozlerhakan** [comment]
Ah well, let's step back first to see how we can use our own casting feature by looking at the current casting class over here https://github.com/ozlerhakan/poiji/blob/master/src/main/java/com/poiji/config/DefaultCasting.java#L237.   As you can see there, we look at the data type of each field type (here https://github.com/ozlerhakan/poiji/blob/master/src/main/java/com/poiji/config/DefaultCasting.java#L250). What you can do is once you find that the field type is boolean, you can convert the value as:

```
    @Test
    public void shouldDisableCellFormatFromAnnotation() {
        PoijiOptions options = PoijiOptions.PoijiOptionsBuilder.settings().withCasting(new MyConfigXLSX()).build();

        List<TestSampleInfo> result = Poiji.fromExcel(new File("src/test/resources/booleans.xlsx"), TestSampleInfo.class, options);

    }

    class TestSampleInfo {
        @ExcelCell(0)
        private Boolean answer;
    }

    static class MyConfigXLSX implements Casting {
        @Override
        public Object castValue(Field fieldType, String value, int row, int column, PoijiOptions options) {
            if (value.equals("yes") && fieldType.getType() == Boolean.class) {
                return Boolean.TRUE;
            }
            return Boolean.FALSE;
        }
    }
```

an excel file like:

```
Column A
yes
no
```

**Reply by ozlerhakan** [reply]
let me know if It helps

**Reply by mrgpd-pfevre** [reply]
Great Thanks for your help.
I had to go a bit further because I wanted to Override the parsing only for some Column.
So I did this :-)
```class MyConfigXLSX extends DefaultCasting implements Casting {
    @Override
    public Object castValue(Field fieldType, String value, int row, int column, PoijiOptions options) {
        if (fieldType.getType() == Boolean.class) {
            if ("true".equalsIgnoreCase(value.trim()) ||
                    "oui".equalsIgnoreCase(value.trim()) ||
                    "yes".equalsIgnoreCase(value.trim()) ||
                    "1".equals(value.trim())) {
                return Boolean.TRUE;
            } else {
                if ("false".equalsIgnoreCase(value.trim()) ||
                        "non".equalsIgnoreCase(value.trim()) ||
                        "no".equalsIgnoreCase(value.trim()) ||
                        "0".equals(value.trim())) {
                    return Boolean.FALSE;
                }
            }
            return null;
            //throw new BooleanParser.BooleanParseException(value);
        }
        return super.castValue(fieldType, value, row, column, options);
    }
```

**Reply by mrgpd-pfevre** [reply]
Indeed I would have override com.poiji.parser.BooleanParser if possible :-)

**Reply by ozlerhakan** [reply]
Cool @mrgpd-pfevre please mark this as answer so that It would help others ☺️

**Comment by ozlerhakan** [comment, marked as answer]
hi @mrgpd-pfevre , could you please mark the discussion as answer? :) thanks!
